egs/librispeech/asr/nnlm/main.py

```diff
@@ -0,0 +1,331 @@
+    try:
+        data_padded = torch.stack(data_padded).to(device).transpose(0, 1)
+        target = torch.flatten(torch.stack(target).to(device).transpose(0, 1))
+        return data_padded, target
+    except:
+        import pdb
+        pdb.set_trace()
+
+    #return torch.stack(data_padded).to(device), torch.stack(target).to(device)
+
+
+def evaluate(data_source):
+    # Turn on evaluation mode which disables dropout.
+    model.eval()
+    total_loss = 0.
+    ntokens = len(corpus.dictionary)
+    if args.model != 'Transformer':
+        hidden = model.init_hidden(eval_batch_size)
+    with torch.no_grad():
+        for i in range(0, len(data_source) // eval_batch_size - 1):
+            data, targets = get_batch(data_source, i)
+            if args.model == 'Transformer':
+                output = model(data)
+                output = output.view(-1, ntokens)
+            else:
+                output, hidden = model(data, hidden)
+                hidden = repackage_hidden(hidden)
+            total_loss += len(data) * criterion(output, targets).item()
+    return total_loss / (len(data_source) - 1)
+
+
+def train():
+    # Turn on training mode which enables dropout.
+    batch_size = args.batch_size
+    model.train()
+    total_loss = 0.
+    start_time = time.time()
+    ntokens = len(corpus.dictionary)
+    if args.model != 'Transformer':
+        hidden = model.init_hidden(batch_size)
+    for batch_idx in range(0, len(train_data) // batch_size - 1):
+        data, targets = get_batch(train_data, batch_idx)
+        # Starting each batch, we detach the hidden state from how it was previously produced.
+        # If we didn't, the model would try backpropagating all the way to start of the dataset.
+        model.zero_grad()
+        if args.model == 'Transformer':
+            output = model(data)
+            output = output.view(-1, ntokens)
+        else:
+            hidden = repackage_hidden(hidden)
+            output, hidden = model(data, hidden)
+
+        # import pdb; pdb.set_trace()
+        loss = criterion(output, targets)
+        loss.backward()
+
+        # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
+        torch.nn.utils.clip_grad_norm_(model.parameters(), args.clip)
+        for p in model.parameters():
+            p.data.add_(p.grad, alpha=-lr)
+
+        #import pdb; pdb.set_trace()
+        total_loss += loss.item()
+
+        if batch_idx % args.log_interval == 0 and batch_idx > 0:
+            cur_loss = total_loss / args.log_interval
+            elapsed = time.time() - start_time
+            print(
+                '| epoch {:3d} | {:5d}/{:5d} batches | lr {:02.2f} | ms/batch {:5.2f} | '
+                'loss {:5.2f} | ppl {:8.2f}'.format(
+                    epoch, batch_idx,
+                    len(train_data) // batch_size, lr,
+                    elapsed * 1000 / args.log_interval, cur_loss,
+                    math.exp(cur_loss)))
```

Comment:
> These perplexities, are they per word or per token?

@danpovey 
The perplexities are computed as `exp(NLL)` and the modelling units are tokens so
PPL is computed with respect to tokens.